_posts/2021-04-28-#17331.md

```diff
@@ -0,0 +1,111 @@
+  We can preempt the cost added by the input by considering each UTXO at its
+  _effective value_, its value reduced by the UTXO's input cost at the given
+  feerate.
+
+- Using _effective values_ in coin selection means that we can freely add and
+  remove inputs until we have have raised enough funds to pay for the
+  transaction's fixed costs: the payment amounts, the transaction header and the
+  output data. Whatever final input set we settle for, the increased transaction
+  size and input costs have already been accounted for.
+
+- The _knapsack solver_ sorts the available UTXOs by effective value, and then
+  combines candidate input sets by randomly selecting or skipping inputs until
+  it has sufficient funds for the transaction. When it finds a viable solution,
+  it keeps the solution if it produces a smaller change than the prior stored
+  solution. Then, the algorithm deselects the last input and continues
+  traversing the UTXO list as described, trying to converge on the smallest
+  change output above `MIN_CHANGE`. This random process is repeated up to 1,000
+  times.
+
+- The _Branch and Bound solver (BnB)_ deterministically searches the complete
+  combination space. It performs a depth-first search on a binary tree where
+  each branch represents the inclusion or exclusion of a UTXO. It explores each
+  inclusion branch first, and backtracks whenever a subtree cannot yield a
+  solution. It returns the first discovered input set that is an "exact match"
+  for the funding requirement of the transaction. A resulting transaction avoids
+  the creation of a change output, but the BnB algorithm is not guaranteed to
+  find a solution even when there are sufficient funds.
+
+- An
+  [OutputGroup](https://github.com/bitcoin-core-review-club/bitcoin/blob/4ac1adda9914d845aaea5804af4801ffec53c701/src/wallet/coinselection.h#L72)
+  collects all UTXOs that were sent to the same public key.
+
+- This PR implements the use of _effective values_ across all different coin
+  selection strategies. Prior, it was only used for the BnB selection where it
+  facilitated the search to be effectively performed. Introducing _effective
+  values_ into the knapsack solver allows some of the pre-selection steps to be
+  shared between BnB and knapsack, and prevents that the knapsack solver needs
+  to be repeated when the resulting input set exceeds the previously allocated
+  fee budget.
+
+## Questions
+
+1. How does the _effective value_ differ from the `m_value` of a UTXO?
+
+2. What is `not_input_fees`? What does this help with?
+
+3. What happens when `m_subtract_fee_outputs` is true?
+
+4. What does `cost_of_change` represent?
+
+5. How do we know the `cost_of_change` in advance?
+
+6. What are `discard_feerate`, `long_term_feerate`, and `effective_feerate`?
+
+7. Why do the `OutputGroups` get calculated separately for BnB and Knapsack?
+
+8. What purpose did the while-loop in `CreateTransaction()` serve that gets removed?
```

Comment:
All of these need links to the code

Reply:
Gonna push my current iteration, then add links.